File: 2025/day01/main.go
package main

import (
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

func main() {
	var input []byte
	var err error

	stat, _ := os.Stdin.Stat()
	if (stat.Mode() & os.ModeCharDevice) == 0 {
		input, err = io.ReadAll(os.Stdin)
		if err != nil {
			panic(err)
		}
	} else {
		input, err = os.ReadFile("input.txt")
		if err != nil {
			panic(err)
		}
	}

	lines := strings.Split(strings.TrimSpace(string(input)), "\n")

	fmt.Println("part 1:", part1(lines))
	fmt.Println("part 2:", part2(lines))
}

func part1(lines []string) int {
	position := 50
	count := 0

	for _, line := range lines {
		if line == "" {
			continue
		}

		direction := line[0]
		distance, err := strconv.Atoi(line[1:])
		if err != nil {
			continue
		}

		// rotate the dial based on direction
		// L = left (counterclockwise, toward lower numbers)
		// R = right (clockwise, toward higher numbers)
		if direction == 'L' {
			position = (position - distance) % 100
			// handle wraparound for negative values (e.g., 5 - 10 = -5 -> 95)
			if position < 0 {
				position = position + 100
			}
		} else if direction == 'R' {
			position = (position + distance) % 100
		}

		// count every time dial points at 0
		if position == 0 {
			count++
		}
	}

	return count
}

func part2(lines []string) int {
	position := 50
	count := 0

	for _, line := range lines {
		if line == "" {
			continue
		}

		direction := line[0]
		distance, err := strconv.Atoi(line[1:])
		if err != nil {
			continue
		}

		// count how many times the dial passes through 0 during rotation
		if direction == 'L' {
			// rotating left (counterclockwise)
			if position == 0 {
				// starting at 0, dial hits 0 again only after complete rotations
				count += distance / 100
			} else {
				// count complete rotations plus one more if the dial reaches 0 in a partial rotation
				count += distance / 100
				if distance%100 >= position {
					count++
				}
			}
			// update position
			position = (position - distance) % 100
			if position < 0 {
				position = position + 100
			}
		} else if direction == 'R' {
			// rotating right (clockwise)
			// count how many times dial crosses 0 boundary
			count += (position + distance) / 100
			// update position
			position = (position + distance) % 100
		}
	}

	return count
}

File: 2025/day01/main.zig
const std = @import("std");

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const input = try readInput(allocator);
    defer allocator.free(input);

    const stdout = std.io.getStdOut().writer();
    try stdout.print("part 1: {d}\n", .{part1(input)});
    try stdout.print("part 2: {d}\n", .{part2(input)});
}

fn readInput(allocator: std.mem.Allocator) ![]const u8 {
    const stdin = std.io.getStdIn();
    const stat = stdin.stat() catch {
        return readFile(allocator, "input.txt");
    };

    if (stat.kind == .character_device) {
        return readFile(allocator, "input.txt");
    }

    return try stdin.readToEndAlloc(allocator, std.math.maxInt(usize));
}

fn readFile(allocator: std.mem.Allocator, filename: []const u8) ![]const u8 {
    const file = std.fs.cwd().openFile(filename, .{}) catch |err| {
        std.debug.print("error: cannot open {s}\n", .{filename});
        return err;
    };
    defer file.close();
    return try file.readToEndAlloc(allocator, std.math.maxInt(usize));
}

fn part1(input: []const u8) i32 {
    var position: i32 = 50;
    var count: i32 = 0;

    var lines = std.mem.splitScalar(u8, input, '\n');
    while (lines.next()) |line| {
        if (line.len == 0) continue;

        const direction = line[0];
        const distance = std.fmt.parseInt(i32, line[1..], 10) catch continue;

        // rotate the dial, L = left (counterclockwise), R = right (clockwise)
        if (direction == 'L') {
            position = @mod(position - distance, 100);
        } else if (direction == 'R') {
            position = @mod(position + distance, 100);
        }

        if (position == 0) {
            count += 1;
        }
    }

    return count;
}

fn part2(input: []const u8) i32 {
    var position: i32 = 50;
    var count: i32 = 0;

    var lines = std.mem.splitScalar(u8, input, '\n');
    while (lines.next()) |line| {
        if (line.len == 0) continue;

        const direction = line[0];
        const distance = std.fmt.parseInt(i32, line[1..], 10) catch continue;

        // count how many times the dial passes through 0 during rotation
        if (direction == 'L') {
            if (position == 0) {
                count += @divTrunc(distance, 100);
            } else {
                count += @divTrunc(distance, 100);
                if (@mod(distance, 100) >= position) {
                    count += 1;
                }
            }
            position = @mod(position - distance, 100);
        } else if (direction == 'R') {
            count += @divTrunc(position + distance, 100);
            position = @mod(position + distance, 100);
        }
    }

    return count;
}
